src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,177 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> first;
+    private Node<T> last;
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> prev;
+        private Node<T> next;
+
+        private Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        addBottom(value);
     }
 
     @Override
     public void add(T value, int index) {
+        checkRangeAdd(index);
+        if (index == 0) {
+            addTop(value);
+        } else if (index == size) {
+            addBottom(value);
+        } else {
+            addMiddle(value, index);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            addBottom(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkRange(index);
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkRange(index);
+        Node<T> nodeAtIndex = getNode(index);
+        T oldValue = nodeAtIndex.value;
+        nodeAtIndex.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkRange(index);
+        Node<T> toRemove = getNode(index);
+        Node<T> nodeBefore = toRemove.prev;
+        Node<T> nodeAfter = toRemove.next;
+        if (nodeBefore == null) {
+            first = nodeAfter;
+        }
+        if (nodeAfter == null) {
+            last = nodeBefore;
+        } else {
+            nodeAfter.prev = nodeBefore;
+        }
+        size--;
```

Comment:
please create  a separate method unlink(Node node) for this code. you already have the method removeNode(). https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html#create-a-separate-method-unlinknode-node